### PR TITLE
[api-spec] query api draft felix

### PR DESCRIPTION
**Pull request by felixhaeberle**
hi folks, this is my idea of a dead simple API.

this PR focuses on the return type to always be an array of messages (empty, one message, multiple messages) and to have the least amount of functions we need.

I also see the value in having more / distinct functions, as it's currently discussed in other PR's but this idea really wants to focus on the basics (we can also make the API bigger afterward).

I skipped the error/exception handling for now, but we need that in this approach because queries could fail if not executed correctly with the right properties. -> ivan corrected me https://github.com/inlang/inlang/pull/1158#issuecomment-1645586360

### CODE REVIEW

**Comment by ivanhofer** [comment]
> I skipped the error/exception handling for now, but we need that in this approach because queries could fail if not executed correctly with the right properties.

I think qeuries should not fail.. they could return an empty array if the input is invalid?

File: source-code/core2/src/messages/query.ts [hunk 1]
```diff
@@ -1,15 +1,48 @@
...
+const messageOrMessages = inlang.messages.get({
+	where: { id: "myMessageId", languageTag: "en" },
+}) as ReturnType<MessagesQuery["get"]>
+
+// - update
+const messageOrMessagesUpdated = inlang.messages.update([
+	{ id: "myMessageId", languageTag: "en", pattern: [{ type: "Text", value: "Hello World" }] },
+]) as ReturnType<MessagesQuery["update"]>
+
+// - delete
```
Comment (NiklasBuchfink) [inline source-code/core2/src/messages/query.ts:39]: I don't think that update to an undefined pattern is the same as a dedicated delete. This leads to confusion.

File: source-code/core2/src/messages/query.ts [hunk 2]
```diff
@@ -1,15 +1,48 @@
...
-// query.
+export type MessagesQuery = {
+	get: (args: QueryOptions) => Message[]
+	update: (data: Message[]) => Message[]
+	list: () => Message[]
+}
+
+/* --------------------- CRUD --------------------- */
+
+// - create
```
Comment (NiklasBuchfink) [inline source-code/core2/src/messages/query.ts:24]: So the `update` function is basically an `upsert`, right?

File: source-code/core2/src/messages/query.ts (hunk 1 above)
Comment (felixhaeberle) [reply source-code/core2/src/messages/query.ts:39]: I think having to decide and flesh out the cases between 5+ crud functions in apps are the bigger burden as passing `undefined` to one function and let core handle the rest.

**Comment by samuelstroschein** [comment]
@felixhaeberle the types are all `any`. I thought of a PR as implementing the API for the query. 

We don't know whether your proposal works from a type perspective.

**Comment by felixhaeberle** [comment]
@samuelstroschein fixed the types 👍

File: source-code/core2/src/messages/query.ts [hunk 3]
```diff
@@ -1,15 +1,51 @@
...
+/* --------------------- TYPES --------------------- */
+
+type FilterType = {
+	id: string
+	languageTag: string
+	// more properties in the future
+}
+
+type QueryOptions = {
+	where: Partial<FilterType>[]
```
Comment (ivanhofer) [inline source-code/core2/src/messages/query.ts:12]: Why an array?
Why a nested object with a single prop `where`?

**Comment by samuelstroschein** [comment]
I think @NilsJacobsen proposal is better :P 

Close this one?

**Comment by felixhaeberle** [comment]
@samuelstroschein No I think it's not, please argument.

**Comment by samuelstroschein** [comment]
> @samuelstroschein No I think it's not, please argument.

Your API is simpler but will eventually break. 

1. Give CRUD ops. Not only `get` and `update` where `update` behaves like an `upsert`. Wtf?
2. Get should return a single message. Hence `get` and `getMany` is better. 

3. The benefit of your API, that you can query multiple messages at once can easily be ported to @NilsJacobsen API: 

```ts

inlang.messages.get({ where: { id: ["myMessage", "anotherMessage"] })
inlang.messages.delete({ where: { id: ["myMessage", "anotherMessage"] }

```

I'm on a call with Nils rn and we are both not sure on how to compare those approaches.

File: source-code/core2/src/messages/query.ts (hunk 3 above)
Comment (ivanhofer) [reply source-code/core2/src/messages/query.ts:12]: is this a use-case?

**Comment by felixhaeberle** [comment]
> Give CRUD ops. Not only get and update where update behaves like an upsert. Wtf?

Fair enough, should have been `get` & `set` 

> Get should return a single message. Hence get and getMany is better.

The goal of this PR was to have a return type of `Message[]` all the time, you are right, there could be situations where I only want the object – without wrapped in an array. 

> The benefit of your API, that you can query multiple messages at once can easily be ported to @NilsJacobsen API:

Yep 👍

**Comment by samuelstroschein** [comment]
Closing in favor of https://github.com/inlang/inlang/pull/1171

**Pull request closed without merging**